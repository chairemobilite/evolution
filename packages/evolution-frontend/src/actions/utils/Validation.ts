import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { ValidationFunction } from 'evolution-common/lib/services/widgets';
import { I18nData } from 'evolution-common/lib/utils/helpers';

/**
 * Check the validations for a specific value
 * @param validations The validation functions or undefined if no function
 * @param value The value of the widget
 * @param customValue Any custom value associated with this widget
 * @param interview The interview data
 * @param path The path of the current widget to validate
 * @param customPath The custom path
 * @returns An array where the first value is a boolean indicating if the value
 * is valid and the second value is the error message if the value is invalid
 */
export const checkValidations = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    validations: ValidationFunction<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> | undefined,
    value: unknown | undefined,
    customValue: unknown | undefined,
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>,
    path: string,
    customPath?: string
): [boolean, I18nData<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> | undefined] => {
    if (typeof validations === 'function') {
        try {
            const validationsGroup = validations(value, customValue, interview, path, customPath);
            for (let i = 0; i < validationsGroup.length; i++) {
                if (validationsGroup[i].validation === true) {
                    return [false, validationsGroup[i].errorMessage];
                }
            }
        } catch (error) {
            // If there is an error during validations, just ignore the error and consider the value as valid to make sure we can at least get the next answers without blocking the questionnaire.
            // TODO: add a server-side log of the error and define a better way to deal with errors.
            console.log('validation error', error);
            return [true, undefined];
        }
    }
    return [true, undefined];
};

export const validateAllWidgets = function <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>
) {
    // TODO Move the allWidgetsValid out of the interview type
    return (interview as any).allWidgetsValid;
};
