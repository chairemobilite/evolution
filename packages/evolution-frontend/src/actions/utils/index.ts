export * from './Conditional';
export * from './Validation';
export * from './WidgetOperation';
