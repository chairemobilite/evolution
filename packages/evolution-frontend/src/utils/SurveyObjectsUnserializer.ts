/*
 * Copyright 2025, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

import { SurveyObjectsWithAudits, AuditForObject, AuditsByObject } from 'evolution-common/lib/services/audits/types';
import { Home } from 'evolution-common/lib/services/baseObjects/Home';
import { Household } from 'evolution-common/lib/services/baseObjects/Household';
import { Interview } from 'evolution-common/lib/services/baseObjects/interview/Interview';
import { SurveyObjectsRegistry } from 'evolution-common/lib/services/baseObjects/SurveyObjectsRegistry';

/**
 * Utility class to unserialize survey objects from their serialized form
 * This is used when receiving surveyObjectsAndAudits from the backend
 */
export class SurveyObjectsUnserializer {
    /**
     * Unserialize surveyObjectsAndAudits from the backend
     * @param serializedData - The serialized SurveyObjectsWithAudits data
     * @returns Unserialized SurveyObjectsWithAudits with proper object instances
     */
    static unserialize(serializedData: unknown): SurveyObjectsWithAudits | undefined {
        if (!serializedData) {
            return undefined;
        }

        // clear first, especially when dealing with mulitple interviews,
        // otherwise the registry will fill out with previous interview objects
        const registry = new SurveyObjectsRegistry();

        try {
            const data = serializedData as Record<string, unknown>;
            const result: SurveyObjectsWithAudits = {
                audits: (data.audits as AuditForObject[]) || [],
                auditsByObject: (data.auditsByObject as AuditsByObject) || {
                    interview: [],
                    household: [],
                    home: [],
                    persons: {},
                    journeys: {},
                    visitedPlaces: {},
                    trips: {},
                    segments: {}
                },
                interview: undefined,
                household: undefined,
                home: undefined
            };

            // Unserialize Interview - now using Interview.unserialize method
            if (data.interview) {
                try {
                    const interviewData = data.interview as Record<string, unknown>;
                    result.interview = Interview.unserialize(interviewData, registry);
                } catch (error) {
                    console.error('Failed to unserialize interview:', error);
                }
            }

            // Unserialize Home - this will handle any nested Place objects
            if (data.home) {
                try {
                    result.home = Home.unserialize(data.home as Record<string, unknown>, registry);
                } catch (error) {
                    console.error('Failed to unserialize home:', error);
                }
            }

            // Unserialize Household - this will automatically handle all nested objects:
            // - members (persons) -> journeys -> visitedPlaces, trips -> segments
            if (data.household) {
                try {
                    const householdData = data.household as Record<string, unknown>;
                    result.household = Household.unserialize(householdData, registry);
                } catch (error) {
                    console.error('Failed to unserialize household:', error);
                }
            }

            return result;
        } catch (error) {
            console.error('Failed to unserialize surveyObjectsAndAudits:', error);
            return undefined;
        }
    }

    /**
     * Check if surveyObjectsAndAudits data is present and valid
     * @param data - The data to check
     * @returns boolean indicating if data is present
     */
    static hasValidData(data: unknown): boolean {
        if (!data || typeof data !== 'object') {
            return false;
        }
        const obj = data as Record<string, unknown>;
        return !!(obj.interview || obj.household || obj.home || (Array.isArray(obj.audits) && obj.audits.length > 0));
    }

    /**
     * Set up parent-child UUID relationships for unserialized objects
     * Note: Objects register themselves in the registry via their constructors
     * @param surveyObjects - The unserialized survey objects
     */
    private static registerObjectsInRegistry(surveyObjects: SurveyObjectsWithAudits): void {
        // Set up household parent UUIDs
        if (surveyObjects.household) {
            if (surveyObjects.interview) {
                surveyObjects.household.interviewUuid = surveyObjects.interview._uuid;
            }
            if (surveyObjects.home) {
                surveyObjects.household.homeUuid = surveyObjects.home._uuid;
            }

            // Set up person parent UUIDs
            if (surveyObjects.household.members) {
                for (const person of surveyObjects.household.members) {
                    person.householdUuid = surveyObjects.household._uuid;

                    // Set up journey parent UUIDs
                    if (person.journeys) {
                        for (const journey of person.journeys) {
                            journey.personUuid = person._uuid;

                            // Set up visited place parent UUIDs
                            if (journey.visitedPlaces) {
                                for (const visitedPlace of journey.visitedPlaces) {
                                    visitedPlace.journeyUuid = journey._uuid;
                                }
                            }

                            // Set up trip parent UUIDs
                            if (journey.trips) {
                                for (const trip of journey.trips) {
                                    trip.journeyUuid = journey._uuid;

                                    // Set up segment parent UUIDs
                                    if (trip.segments) {
                                        for (const segment of trip.segments) {
                                            segment.tripUuid = trip._uuid;
                                        }
                                    }

                                    // Set up junction parent UUIDs
                                    if (trip.junctions) {
                                        for (const junction of trip.junctions) {
                                            junction.tripUuid = trip._uuid;
                                        }
                                    }
                                }
                            }

                            // Set up trip chain parent UUIDs
                            if (journey.tripChains) {
                                for (const tripChain of journey.tripChains) {
                                    tripChain.journeyUuid = journey._uuid;
                                }
                            }
                        }
                    }

                    // Set up vehicle parent UUIDs
                    if (person.vehicles) {
                        for (const vehicle of person.vehicles) {
                            vehicle.ownerUuid = person._uuid;
                        }
                    }
                }
            }
        }
    }
}
