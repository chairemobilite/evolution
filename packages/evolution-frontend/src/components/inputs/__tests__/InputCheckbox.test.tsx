/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { faCrow } from '@fortawesome/free-solid-svg-icons/faCrow';

import { shuffle } from 'chaire-lib-common/lib/utils/RandomUtils';
import { interviewAttributes } from './interviewData.test';

import InputCheckbox from '../InputCheckbox';
import i18next from 'i18next';

jest.mock('chaire-lib-common/lib/utils/RandomUtils', () => ({
    shuffle: jest.fn()
}));

const shuffleMock = shuffle as jest.MockedFunction<typeof shuffle>;

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

describe('Render InputCheckbox with various parameter combinations, all parameters', () => {

    const conditionalFct = jest.fn().mockReturnValue(true);
    const translationFct = jest.fn().mockReturnValue('Translated string');
    const choices = [
        {
            value: 'val1',
            label: { en: 'english value', fr: 'valeur française' },
            hidden: false,
            icon: faCrow
        },
        {
            value: 'val2',
            label: 'Unilingual label',
            iconPath: 'img/test.png'
        },
        {
            value: 'hiddenVal',
            label: { en: 'english hidden', fr: 'cachée' },
            hidden: true
        },
        {
            value: 'conditionalVal',
            label: { en: 'english conditional', fr: 'conditionnelle' },
            conditional: conditionalFct
        },
        {
            value: 'val3',
            label: translationFct,
            hidden: false,
            color: 'green'
        }
    ];

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'checkbox' as const,
        choices,
        iconSize: '3em',
        seed: 23,
        shuffle: false,
        addCustom: false,
        columns: 2,
        sameLine: false,
        customLabel: 'custom label',
        size: 'medium' as const,
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    };

    beforeEach(() => {
        conditionalFct.mockClear();
    })

    test('Includes hidden values, conditional displayed, no custom, 2 columns', () => {
        const wrapper = TestRenderer.create(
            <InputCheckbox
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value='value'
                inputRef={React.createRef()}
                
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
        expect(conditionalFct).toHaveBeenCalledTimes(1);
        expect(conditionalFct).toHaveBeenCalledWith(interviewAttributes, 'foo.test', userAttributes);
        expect(translationFct).toHaveBeenCalledWith(i18next.t, interviewAttributes, 'foo.test', userAttributes);
    });

    test('Conditional value hidden, same line', () => {
        conditionalFct.mockReturnValueOnce(false);
        const testWidgetConfig = Object.assign({}, widgetConfig, { sameLine: true, columns: undefined });
        const wrapper = TestRenderer.create(
            <InputCheckbox
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={testWidgetConfig}
                value='value'
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Shuffle choices', () => {
        shuffleMock.mockReturnValue([choices[1], choices[3], choices[0], choices[2]]);
        const testWidgetConfig = Object.assign({}, widgetConfig, { shuffle: true });
        const wrapper = TestRenderer.create(
            <InputCheckbox
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={testWidgetConfig}
                value='value'
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
        expect(shuffleMock).toHaveBeenCalledTimes(1);
        expect(shuffleMock).toHaveBeenCalledWith(choices, undefined, widgetConfig.seed);
    });

    test('With custom choice', () => {
        const testWidgetConfig = Object.assign({}, widgetConfig, { addCustom: true });
        const wrapper = TestRenderer.create(
            <InputCheckbox
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={testWidgetConfig}
                value='value'
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
                customId='foo.test.custom'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
    
});

describe('Render InputCheckbox with minimum parameters', () => {

    const choices = [
        {
            value: 'val1',
            label: { en: 'english value', fr: 'valeur française' }
        },
        {
            value: 'val2',
            label: 'Unilingual label'
        }
    ];

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'checkbox' as const,
        choices,
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    };

    test('Minimum parameters', () => {
        const wrapper = TestRenderer.create(
            <InputCheckbox
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value='value'
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
    
});

describe('Render InputCheckbox with HTML label', () => {

    const choices = [
        {
            value: 'first value',
            label: {
                fr: '<div>premiere valeur</div>',
                en: '<div>first value</div>'
            }
        },
        {
            value: 'second value',
            label: '<div>second value</div>'
        }
    ];

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'checkbox' as const,
        choices,
        size: 'medium',
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    };

    test('HTML label', () => {
        const wrapper = TestRenderer.create(
            <InputCheckbox
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value='value'
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

});
