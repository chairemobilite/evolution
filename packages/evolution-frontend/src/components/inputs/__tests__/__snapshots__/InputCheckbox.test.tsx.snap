// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render InputCheckbox with HTML label HTML label 1`] = `
<div
  className="survey-question__input-checkbox-group-container"
>
  <div
    className="survey-question__input-checkbox-input-container"
    onClick={[Function]}
  >
    <div
      className="label-input-container"
    >
      <input
        checked={false}
        className="input-checkbox"
        id="test_first value__input-checkbox__first value"
        name="test_first value"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
        value="first value"
      />
      <label
        htmlFor="test_first value__input-checkbox__first value"
        onClick={[Function]}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "<div>first value</div>",
            }
          }
        />
      </label>
    </div>
  </div>
  <div
    className="survey-question__input-checkbox-input-container"
    onClick={[Function]}
  >
    <div
      className="label-input-container"
    >
      <input
        checked={false}
        className="input-checkbox"
        id="test_second value__input-checkbox__second value"
        name="test_second value"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
        value="second value"
      />
      <label
        htmlFor="test_second value__input-checkbox__second value"
        onClick={[Function]}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "<div>second value</div>",
            }
          }
        />
      </label>
    </div>
  </div>
</div>
`;

exports[`Render InputCheckbox with minimum parameters Minimum parameters 1`] = `
<div
  className="survey-question__input-checkbox-group-container"
>
  <div
    className="survey-question__input-checkbox-input-container"
    onClick={[Function]}
  >
    <div
      className="label-input-container"
    >
      <input
        checked={false}
        className="input-checkbox"
        id="test_val1__input-checkbox__val1"
        name="test_val1"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
        value="val1"
      />
      <label
        htmlFor="test_val1__input-checkbox__val1"
        onClick={[Function]}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "english value",
            }
          }
        />
      </label>
    </div>
  </div>
  <div
    className="survey-question__input-checkbox-input-container"
    onClick={[Function]}
  >
    <div
      className="label-input-container"
    >
      <input
        checked={false}
        className="input-checkbox"
        id="test_val2__input-checkbox__val2"
        name="test_val2"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
        value="val2"
      />
      <label
        htmlFor="test_val2__input-checkbox__val2"
        onClick={[Function]}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "Unilingual label",
            }
          }
        />
      </label>
    </div>
  </div>
</div>
`;

exports[`Render InputCheckbox with various parameter combinations, all parameters Conditional value hidden, same line 1`] = `
<div
  className="survey-question__input-checkbox-group-container"
>
  <div
    className="survey-question__input-checkbox-input-container"
    onClick={[Function]}
  >
    <div
      className="label-input-container"
    >
      <input
        checked={false}
        className="input-checkbox"
        id="test_val1__input-checkbox__val1"
        name="test_val1"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
        value="val1"
      />
      <label
        htmlFor="test_val1__input-checkbox__val1"
        onClick={[Function]}
      >
        <span>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-crow fa-w-20 faIconLeft"
            data-icon="crow"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={
              Object {
                "margin": "auto 0",
              }
            }
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M544 32h-16.36C513.04 12.68 490.09 0 464 0c-44.18 0-80 35.82-80 80v20.98L12.09 393.57A30.216 30.216 0 0 0 0 417.74c0 22.46 23.64 37.07 43.73 27.03L165.27 384h96.49l44.41 120.1c2.27 6.23 9.15 9.44 15.38 7.17l22.55-8.21c6.23-2.27 9.44-9.15 7.17-15.38L312.94 384H352c1.91 0 3.76-.23 5.66-.29l44.51 120.38c2.27 6.23 9.15 9.44 15.38 7.17l22.55-8.21c6.23-2.27 9.44-9.15 7.17-15.38l-41.24-111.53C485.74 352.8 544 279.26 544 192v-80l96-16c0-35.35-42.98-64-96-64zm-80 72c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "english value",
            }
          }
        />
      </label>
    </div>
  </div>
  <div
    className="survey-question__input-checkbox-input-container"
    onClick={[Function]}
  >
    <div
      className="label-input-container"
    >
      <input
        checked={false}
        className="input-checkbox"
        id="test_val2__input-checkbox__val2"
        name="test_val2"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
        value="val2"
      />
      <label
        htmlFor="test_val2__input-checkbox__val2"
        onClick={[Function]}
      >
        <span>
          <img
            alt="Unilingual label"
            className="faIconLeft"
            src="img/test.png"
            style={
              Object {
                "height": "3em",
                "margin": "0 0.25em 0 0",
              }
            }
          />
        </span>
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "Unilingual label",
            }
          }
        />
      </label>
    </div>
  </div>
  <div
    className="survey-question__input-checkbox-input-container"
    onClick={[Function]}
  >
    <div
      className="label-input-container"
    >
      <input
        checked={false}
        className="input-checkbox"
        id="test_val3__input-checkbox__val3"
        name="test_val3"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
        value="val3"
      />
      <label
        htmlFor="test_val3__input-checkbox__val3"
        onClick={[Function]}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "Translated string",
            }
          }
        />
      </label>
    </div>
  </div>
</div>
`;

exports[`Render InputCheckbox with various parameter combinations, all parameters Includes hidden values, conditional displayed, no custom, 2 columns 1`] = `
<div
  className="survey-question__input-checkbox-group-container no-wrap"
>
  <div
    className="survey-question__input-checkbox-group-column"
  >
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_val1__input-checkbox__val1"
          name="test_val1"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="val1"
        />
        <label
          htmlFor="test_val1__input-checkbox__val1"
          onClick={[Function]}
        >
          <span>
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-crow fa-w-20 faIconLeft"
              data-icon="crow"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={
                Object {
                  "margin": "auto 0",
                }
              }
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M544 32h-16.36C513.04 12.68 490.09 0 464 0c-44.18 0-80 35.82-80 80v20.98L12.09 393.57A30.216 30.216 0 0 0 0 417.74c0 22.46 23.64 37.07 43.73 27.03L165.27 384h96.49l44.41 120.1c2.27 6.23 9.15 9.44 15.38 7.17l22.55-8.21c6.23-2.27 9.44-9.15 7.17-15.38L312.94 384H352c1.91 0 3.76-.23 5.66-.29l44.51 120.38c2.27 6.23 9.15 9.44 15.38 7.17l22.55-8.21c6.23-2.27 9.44-9.15 7.17-15.38l-41.24-111.53C485.74 352.8 544 279.26 544 192v-80l96-16c0-35.35-42.98-64-96-64zm-80 72c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </span>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "english value",
              }
            }
          />
        </label>
      </div>
    </div>
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_val2__input-checkbox__val2"
          name="test_val2"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="val2"
        />
        <label
          htmlFor="test_val2__input-checkbox__val2"
          onClick={[Function]}
        >
          <span>
            <img
              alt="Unilingual label"
              className="faIconLeft"
              src="img/test.png"
              style={
                Object {
                  "height": "3em",
                  "margin": "0 0.25em 0 0",
                }
              }
            />
          </span>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Unilingual label",
              }
            }
          />
        </label>
      </div>
    </div>
  </div>
  <div
    className="survey-question__input-checkbox-group-column"
  >
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_conditionalVal__input-checkbox__conditionalVal"
          name="test_conditionalVal"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="conditionalVal"
        />
        <label
          htmlFor="test_conditionalVal__input-checkbox__conditionalVal"
          onClick={[Function]}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "english conditional",
              }
            }
          />
        </label>
      </div>
    </div>
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_val3__input-checkbox__val3"
          name="test_val3"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="val3"
        />
        <label
          htmlFor="test_val3__input-checkbox__val3"
          onClick={[Function]}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Translated string",
              }
            }
          />
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render InputCheckbox with various parameter combinations, all parameters Shuffle choices 1`] = `
<div
  className="survey-question__input-checkbox-group-container no-wrap"
>
  <div
    className="survey-question__input-checkbox-group-column"
  >
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_val2__input-checkbox__val2"
          name="test_val2"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="val2"
        />
        <label
          htmlFor="test_val2__input-checkbox__val2"
          onClick={[Function]}
        >
          <span>
            <img
              alt="Unilingual label"
              className="faIconLeft"
              src="img/test.png"
              style={
                Object {
                  "height": "3em",
                  "margin": "0 0.25em 0 0",
                }
              }
            />
          </span>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Unilingual label",
              }
            }
          />
        </label>
      </div>
    </div>
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_conditionalVal__input-checkbox__conditionalVal"
          name="test_conditionalVal"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="conditionalVal"
        />
        <label
          htmlFor="test_conditionalVal__input-checkbox__conditionalVal"
          onClick={[Function]}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "english conditional",
              }
            }
          />
        </label>
      </div>
    </div>
  </div>
  <div
    className="survey-question__input-checkbox-group-column"
  >
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_val1__input-checkbox__val1"
          name="test_val1"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="val1"
        />
        <label
          htmlFor="test_val1__input-checkbox__val1"
          onClick={[Function]}
        >
          <span>
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-crow fa-w-20 faIconLeft"
              data-icon="crow"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={
                Object {
                  "margin": "auto 0",
                }
              }
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M544 32h-16.36C513.04 12.68 490.09 0 464 0c-44.18 0-80 35.82-80 80v20.98L12.09 393.57A30.216 30.216 0 0 0 0 417.74c0 22.46 23.64 37.07 43.73 27.03L165.27 384h96.49l44.41 120.1c2.27 6.23 9.15 9.44 15.38 7.17l22.55-8.21c6.23-2.27 9.44-9.15 7.17-15.38L312.94 384H352c1.91 0 3.76-.23 5.66-.29l44.51 120.38c2.27 6.23 9.15 9.44 15.38 7.17l22.55-8.21c6.23-2.27 9.44-9.15 7.17-15.38l-41.24-111.53C485.74 352.8 544 279.26 544 192v-80l96-16c0-35.35-42.98-64-96-64zm-80 72c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </span>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "english value",
              }
            }
          />
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render InputCheckbox with various parameter combinations, all parameters With custom choice 1`] = `
<div
  className="survey-question__input-checkbox-group-container no-wrap"
>
  <div
    className="survey-question__input-checkbox-group-column"
  >
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_val1__input-checkbox__val1"
          name="test_val1"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="val1"
        />
        <label
          htmlFor="test_val1__input-checkbox__val1"
          onClick={[Function]}
        >
          <span>
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-crow fa-w-20 faIconLeft"
              data-icon="crow"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={
                Object {
                  "margin": "auto 0",
                }
              }
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M544 32h-16.36C513.04 12.68 490.09 0 464 0c-44.18 0-80 35.82-80 80v20.98L12.09 393.57A30.216 30.216 0 0 0 0 417.74c0 22.46 23.64 37.07 43.73 27.03L165.27 384h96.49l44.41 120.1c2.27 6.23 9.15 9.44 15.38 7.17l22.55-8.21c6.23-2.27 9.44-9.15 7.17-15.38L312.94 384H352c1.91 0 3.76-.23 5.66-.29l44.51 120.38c2.27 6.23 9.15 9.44 15.38 7.17l22.55-8.21c6.23-2.27 9.44-9.15 7.17-15.38l-41.24-111.53C485.74 352.8 544 279.26 544 192v-80l96-16c0-35.35-42.98-64-96-64zm-80 72c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </span>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "english value",
              }
            }
          />
        </label>
      </div>
    </div>
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_val2__input-checkbox__val2"
          name="test_val2"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="val2"
        />
        <label
          htmlFor="test_val2__input-checkbox__val2"
          onClick={[Function]}
        >
          <span>
            <img
              alt="Unilingual label"
              className="faIconLeft"
              src="img/test.png"
              style={
                Object {
                  "height": "3em",
                  "margin": "0 0.25em 0 0",
                }
              }
            />
          </span>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Unilingual label",
              }
            }
          />
        </label>
      </div>
    </div>
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_conditionalVal__input-checkbox__conditionalVal"
          name="test_conditionalVal"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="conditionalVal"
        />
        <label
          htmlFor="test_conditionalVal__input-checkbox__conditionalVal"
          onClick={[Function]}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "english conditional",
              }
            }
          />
        </label>
      </div>
    </div>
  </div>
  <div
    className="survey-question__input-checkbox-group-column"
  >
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_val3__input-checkbox__val3"
          name="test_val3"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="val3"
        />
        <label
          htmlFor="test_val3__input-checkbox__val3"
          onClick={[Function]}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Translated string",
              }
            }
          />
        </label>
      </div>
    </div>
    <div
      className="survey-question__input-checkbox-input-container"
      onClick={[Function]}
    >
      <div
        className="label-input-container"
      >
        <input
          checked={false}
          className="input-checkbox"
          id="test_custom__input-checkbox__custom"
          name="test_custom"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
          value="custom"
        />
        <label
          htmlFor="test_custom__input-checkbox__custom"
          onClick={[Function]}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Other",
              }
            }
          />
        </label>
      </div>
    </div>
  </div>
  <div
    className="label-input-custom-container"
  >
    <label
      htmlFor="foo.test.custom"
    >
      <span>
        custom label
      </span>
    </label>
    <input
      autoComplete="none"
      className="apptr__form-input input-custom apptr__input-string"
      id="foo.test.custom"
      name="foo.test.custom"
      onBlur={[Function]}
      onChange={[Function]}
      tabIndex={-1}
      type="text"
      value=""
    />
  </div>
</div>
`;
